File: 프로그래머스/4/64063. 호텔 방 배정/호텔 방 배정.py
import sys
sys.setrecursionlimit(10**6)

parent = {}

def find(num):
    
    if num not in parent:
        return num
    
    root = find(parent[num])
    parent[num] = root
    return root

def union(a, b):
    a = find(a)
    b = find(b)
    
    if a != b:
        parent[a] = b

def solution(k, room_number):
    answer = []
    
    for num in room_number:
        found = find(num)
        answer.append(found)
        
        union(num, found + 1)
    
    return answer

File: 프로그래머스/4/64063. 호텔 방 배정/호텔 방 배정.java
import java.util.*;

class Solution {
    
    static Map<Long, Long> parent = new HashMap<>();
    
    public long[] solution(long k, long[] room_number) {
        
        long[] ans = new long[room_number.length];

        for (int i = 0; i < room_number.length; i++) {

            long num = room_number[i];
            long found = find(num);

            ans[i] = found;

            union(num, found + 1);
        }

        return ans;
    }

    public long find(long num) {

        if (!parent.containsKey(num)) {
            return num;
        }

        long root = find(parent.get(num));
        parent.put(num, root);
        return root;
    }

    public void union(long a, long b) {
        a = find(a);
        b = find(b);

        if (a != b) {
            parent.put(a, b);
        }
    }
}